Feat/add uhpo crate rework

jayendramadaram opened the pull request:
This Pr is First pr in series of broken Pr's from #9 

### Changes Made
- Added `payout-update` Transaction builder.
- Offers way to Schnorr Sign `payout-update`  Transactions.
- removed dependency from nigiri from testcase perspective.
- includes unit tests for `payout-update`.

jayendramadaram committed:
feat/added payout update

jayendramadaram committed:
feat/added payout update transaction Builder

pool2win commented on .gitignore:
```diff
@@ -16,4 +18,5 @@ proposal/proposal.blg
-target
\ No newline at end of file
+target
+todo
```
Move this to .git/config/ignore if this is your personal ignore file. Other contributors don't have this file.

pool2win commented on node/connection/src/lib.rs:
```diff
@@ -343,10 +344,7 @@ mod tests {
     #[tokio::test]
-    // TODO: This is a smoke test and we need to improve how we wait for listen to complete before we invoke connect.
     async fn it_should_run_connect_without_errors() {
-        use tokio::time::{sleep, Duration};
```
Why did you remove these? This change doesn't seem related to the UHPO work, so if we need this change to go in, let's make it a separate PR.

pool2win commented on node/connection/src/lib.rs:
```diff
@@ -358,29 +356,34 @@ mod tests {
 
-        // TODO: Fix this smoke test! Kill the sleep in this smoke test.
-        sleep(Duration::from_millis(100)).await;
+        notify.notified().await;
```
I think this change of notifier is valuable, but can you please move this into a separate PR so we can test and review it as a unit.

pool2win commented on node/uhpo/docs/spec.md:
```diff
@@ -0,0 +1,320 @@
+# Braidpool UHPO Spec
```
We should not have this design doc in the repo. I am not sure the design will always remain in sync with the code, and thus will cause confusion. Let's move this into a gist and link the gist from the PR description. This way the history of this discussion is maintained and we are not burdened to keep it in sync with the code as it evolves.

pool2win commented on node/uhpo/src/error.rs:
```diff
@@ -0,0 +1,46 @@
+impl std::error::Error for UhpoError {}
+
+impl UhpoError {
+    pub fn new_other_error(message: &str) -> Self {
```
Why not simply call this method `new`?

pool2win reviewed (changes requested):
@jayendramadaram - I left some initial comments that need to be addressed. I am still going through payout_update.rs and will follow up with another review.

pool2win commented:
I see what happened with the changes for notifier. It seems like there is a problem with this PR where the commits from stutxo have crept into this PR. We need to clear this up, so that only your commits relevant to this PR are in this PR.

pool2win commented on node/uhpo/src/payout_update.rs:
```diff
@@ -0,0 +1,272 @@
+
+impl<'a> PayoutUpdate<'a> {
+    pub fn new(
+        prev_update_tx: Option<&'a Transaction>,
```
I think we can let these arguments can be passed in without using references. I think the client can be required to transfer ownership to the payout update value. If the client wants to retain ownership, then they can Clone the value and pass it in here.

The reason to use ownership transfer instead of references is to keep the code simpler and therefore easier to maintain and use in the future.

Let's try to do this, and I can review again.

jayendramadaram replied:
I've implemented the changes as suggested,
- Removed lifetime parameters from the PayoutUpdate struct.
- Modified the PayoutUpdate struct to take ownership of Transaction objects
- Adjusted internal methods to work with owned TxOut instances

```rust
pub struct PayoutUpdate {
    transaction: Transaction,
    coinbase_txout: TxOut,
    prev_update_txout: Option<TxOut>,
}

impl PayoutUpdate {
    pub fn new(
        prev_update_tx: Option<Transaction>,
        coinbase_tx: Transaction,
        next_out_address: Address,
        projected_fee: Amount,
    ) -> Result<Self, Error> {
        let prev_update_txout = prev_update_tx.as_ref().map(|tx| tx.output[0].clone());
        let coinbase_txout = coinbase_tx.output[0].clone();

        // ... rest of the implementation ...
    }

    // ... other methods updated accordingly ...
}
```

pool2win reviewed (changes requested):
My second review here, about using ownership transfer instead of references with lifetimes.

jayendramadaram committed:
fix/Refactor PayoutUpdate to use ownership transfer and improve crate

jayendramadaram commented:
I have, made above suggested changes, latest commit includes

- removal of personal ignore files from .gitignore
- removal of spec docs
- This PR is rebased to remove commits from stutxo
- Modified the PayoutUpdate struct to take ownership of Transaction objects

pool2win commented on .gitignore:
```diff
@@ -15,5 +15,4 @@ proposal/proposal.bbl
-venv
-target
\ No newline at end of file
+venv
```
The venv should be in your private .git/info/exclude

pool2win commented on node/uhpo/src/payout_update.rs:
```diff
@@ -0,0 +1,275 @@
+}
+
+fn build_transaction(
+    coinbase_tx: &Transaction,
+    prev_update_tx: Option<&Transaction>,
+    next_out_address: Address,
+    projected_fee: Amount,
+    coinbase_txout: &TxOut,
+    prev_update_txout: Option<&TxOut>,
```
We don't need any of these arguments to be a reference.

Here's a working patch that shows how. https://github.com/pool2win/braidpool/commit/638d3c74d6e6695b535f241985ce3de010ed756b

pool2win commented on node/uhpo/src/payout_update.rs:
```diff
@@ -0,0 +1,275 @@
+        total_amount += tx_out.value;
+    }
+
+    let mut payout_update_tx = Transaction {
+        version: Version::TWO,
+        lock_time: LockTime::ZERO,
+        input: vec![],
+        output: vec![],
+    };
+
+    payout_update_tx.input.push(TxIn {
+        previous_output: OutPoint {
+            txid: coinbase_tx.compute_txid(),
+            vout: 0,
+        },
+        script_sig: ScriptBuf::new(),
+        sequence: Sequence::MAX,
+        witness: Witness::default(),
+    });
+
+    if let Some(tx) = prev_update_tx {
+        payout_update_tx.input.push(TxIn {
+            previous_output: OutPoint {
+                txid: tx.compute_txid(),
+                vout: 0,
+            },
+            script_sig: ScriptBuf::new(),
+            sequence: Sequence::MAX,
+            witness: Witness::default(),
+        });
+    }
+
+    payout_update_tx.output.push(TxOut {
+        value: total_amount - projected_fee,
+        script_pubkey: next_out_address.script_pubkey(),
+    });
+
+    Ok(payout_update_tx)
```
There are a lot of individual steps we are taking here. Maybe we should use a builder pattern here again and each of the separate blocks can be turned into separate functions? That will also result in the `build_transaction` function either disappearing or getting simpler.

https://rust-unofficial.github.io/patterns/patterns/creational/builder.html

jayendramadaram replied:
I 'have Implemented changes as suggested,
- added new `TransactionBuilder` which can be used but both `PayoutSettlement` and `PayoutUpdate`

pool2win commented on node/uhpo/src/payout_update.rs:
```diff
@@ -0,0 +1,275 @@
+fn add_signature(
+    transaction: &mut Transaction,
+    input_idx: usize,
+    prevouts: &[&TxOut],
+    private_key: &SecretKey,
+    tweak: &Option<Scalar>,
```
Again, references. My understanding right now with Rust is - use move, it is efficient in Rust, use refs only if really needed.

jayendramadaram replied:
I have changes private_key from not to take as reference, but other two prevouts and tweak are meant to be passed as reference to rust-bitcoin's `SighashCache.taproot_key_spend_signature_hash` and `Keypair.add_xonly_tweak`, hence using  continuing with reference only for these two params.

pool2win commented on node/uhpo/src/payout_update.rs:
```diff
@@ -0,0 +1,275 @@
+    let secp = Secp256k1::new();
+
+    let keypair: Keypair = match tweak {
+        Some(tweak) => private_key.keypair(&secp).add_xonly_tweak(&secp, tweak)?,
```
We should differentiate the error here and the one on line 158, so that the client knows where the signing failed.

jayendramadaram replied:
Added new errors as suggested

- UhpoError::KeypairCreationError
- UhpoError::SignatureVerificationError

pool2win commented on node/uhpo/src/payout_update.rs:
```diff
@@ -0,0 +1,275 @@
+        assert_eq!(payout_update.transaction.input[1].witness.len(), 1);
+    }
+
+    fn create_dummy_transaction() -> Transaction {
```
Move this function to the top, below the setup. All support functions should come before the test functions.

pool2win commented on node/uhpo/src/payout_update.rs:
```diff
@@ -0,0 +1,275 @@
+    }
+
+    #[test]
+    fn test_add_signature() {
```
We should add tests for the code paths where:
1. No tweak is passed
2. There is an error on line 140
3. There is an error on line 158
4. Add test where signature verification fails - you might need to use mocks for this

jayendramadaram replied:
With latest changes made,
- new testcases are added to cover tweak keypair case.
- mock `KeypairCreationError` and `SignatureVerificationError` using `mockall` crate and making SecretKey , Secp256k1 and Keypair Generic.

pool2win commented on node/uhpo/src/payout_update.rs:
```diff
@@ -0,0 +1,275 @@
+    }
+}
+
+fn build_transaction(
```
Add doc comment

pool2win commented on node/uhpo/src/payout_update.rs:
```diff
@@ -0,0 +1,275 @@
+        add_signature(&mut self.transaction, 0, &prevouts, private_key, tweak)
+    }
+
+    pub fn add_prev_update_sig(
```
Add doc comment

pool2win commented on node/uhpo/src/payout_update.rs:
```diff
@@ -0,0 +1,275 @@
+        })
+    }
+
+    pub fn add_coinbase_sig(
```
Add doc comment

pool2win commented on node/uhpo/src/payout_update.rs:
```diff
@@ -0,0 +1,275 @@
+
+use crate::error::UhpoError;
+
+pub struct PayoutUpdate {
```
Add doc comment

pool2win reviewed (changes requested):
Nice progress. A few more things we need to do. See individual comments.

jayendramadaram committed:
feat/add new tx builder

jayendramadaram committed:
test/add mock test for key creation

jayendramadaram committed:
test/add sig verification mock tests and Doc comments

jayendramadaram committed:
test/add sig verification mock tests and Doc comments

jayendramadaram commented:
I have implemented the suggested changes in the latest commit. The updates include:

1. Added  doc strings and comments to code .
2. Implemented a separate `TransactionBuilder` for constructing Bitcoin transactions in `PayoutUpdate`, improving modularity and reusability.
3. Moved `add_signatures` function to `src/crypto/signature.rs`, as it's used by both `PayoutUpdate` and `PayoutSettlement`.
4. Introduced new, more descriptive error types.
5. Implemented mock tests to simulate `rust-bitcoin` instances returning errors.

pool2win commented on node/uhpo/src/crypto/signature.rs:
```diff
@@ -0,0 +1,269 @@
+    }
+
+    #[test]
+    fn test_add_signatures_key_creation_fails() {
```
Let's rename test to 

`test_add_signature_with_invalid_tweak_error_signature_creation_should_fail`

basically

`test_<fn>_<under_which_conditions>_<what_should_happen>`

pool2win commented on node/uhpo/src/crypto/signature.rs:
```diff
@@ -0,0 +1,269 @@
+    }
+
+    #[test]
+    fn test_add_signatures_signature_verification_fails() {
```
fn name here again,

`test_add_signature_with_a_bad_signature_should_fail_during_verification`

pool2win commented on node/uhpo/src/crypto/signature.rs:
```diff
@@ -0,0 +1,269 @@
+}
+
+#[cfg(test)]
+mod mock_tests {
```
Let's add a success case as well. So that we have a unit test for the good case.

pool2win commented on node/uhpo/src/payout_update.rs:
```diff
@@ -0,0 +1,282 @@
+
+// unit tests
+#[cfg(test)]
+mod tests {
```
Let's add a test to see we handle errors properly from `add_signature`. If there is a match on the error, we should have one test case for each.

pool2win reviewed (changes requested):
Nice! I am still not sure if the new approach adds a function call overhead - I'll look at that again.

For now a few things:

1. CI is failing.
2. We should add the failure/success test cases appropriately. I left comments in the PR review for where we need more tests.

jayendramadaram committed:
test/refractor testcase func handles

jayendramadaram commented:
Fixes made from latest commit,
- changed testcase names as requested
- added a mock error testcase which mocks `KeyCreationError(_)` on `add_coinbase_sig` and `add_prev_update_sig` methods on `PayoutUpdate`. I feel single mock testcase should be fine, ignoring testcase for `SignatureVerificationError`.
